### PR TITLE
Fix overflow cases of `ErrorReturnDataOutOfBound`

### PR DESCRIPTION
**Pull request by silathdiir**
### Description

Fix overflow cases of `ErrorReturnDataOutOfBound` state. Found by `testool` cases (`bufferSrcOffset_d108(fail)_g0_v0`).

### Type of change

- [X] Bug fix (non-breaking change which fixes an issue)

### How Has This Been Tested?

Refactor unit-test cases of `ErrorReturnDataOutOfBound` to 4 categories:
- data_offset > u64::MAX
- data_offset + size > U256::MAX
- data_offset + size (mod U256) > u64::MAX
- data_offset + size > return_data_length

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs [hunk 1]
```diff
@@ -84,13 +83,23 @@ impl<F: Field> ExecutionGadget<F> for ErrorReturnDataOutOfBoundGadget<F> {
...
+        // offset plus length may exceed a Word and remainder (end) may be less
+        // than `u64::MAX` (e.g. Word::MAX + 1).
+        cb.require_equal(
+            "Any of [offset > u64::MAX, end > U256::MAX, remainder_end > u64::MAX, remainder_end > return_data_length] occurs",
+            or::expr([
+                // offset > u64::MAX
+                not::expr(is_data_offset_within_range.expr()),
+                // end > U256::MAX
+                sum.carry().as_ref().unwrap().expr(),
+                // remainder_end > u64::MAX
```
Comment (hero78119) [inline zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs:97]: how about give `sum.carry().as_ref().unwrap().expr()` an intermediate variable `is_end_no_overflow`. In this case, maybe rename other `end` with `remainder_end` to match with naming in coding

File: zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs [hunk 2]
```diff
@@ -84,13 +83,23 @@ impl<F: Field> ExecutionGadget<F> for ErrorReturnDataOutOfBoundGadget<F> {
...
-        cb.require_in_set(
-            "Any of [offset_out_of_range, end_out_of_range, end_exceed_length] occurs",
-            not::expr(is_data_offset_within_range.expr())
-                + not::expr(is_end_within_range.expr())
-                + is_end_exceed_length.expr(),
-            vec![1.expr(), 2.expr(), 3.expr()],
+
+        // `end > U256::MAX` is necessary to check with AddWordsGadget carry. Since
+        // offset plus length may exceed a Word and remainder (end) may be less
+        // than `u64::MAX` (e.g. Word::MAX + 1).
```
Comment (hero78119) [inline zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs:89]: How about comment change as below
```rust
// Need to assure `data + offset == reminder_end` in all situation
// once above satisfied then checking `reminder_end` is equivalent constraint `data + offset`. 
// To assure that, `AddWordsGadget carry` must no overflow.
```

File: zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs [hunk 3]
```diff
@@ -84,13 +83,23 @@ impl<F: Field> ExecutionGadget<F> for ErrorReturnDataOutOfBoundGadget<F> {
...
+        cb.require_equal(
+            "Any of [offset > u64::MAX, end > U256::MAX, remainder_end > u64::MAX, remainder_end > return_data_length] occurs",
+            or::expr([
+                // offset > u64::MAX
+                not::expr(is_data_offset_within_range.expr()),
+                // end > U256::MAX
```
Comment (hero78119) [inline zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs:95]: `data + offset > U256::MAX` seems more self-explained

**Review by hero78119 — APPROVED** [review]
LGTM and approved! 
just have comments suggestion hopefully make the logic more self-explanation

File: zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs (hunk 1 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs:95]: Added a new variable in commit [4ea268b](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1321/commits/4ea268b414ea757e452eab267e9a8a656c0cadcc#diff-2f2120439adc744a42a4ca326eedbd453e6b8321cce2105a2e64741e17b67cedR74) (if `carry == 1`, `data_offset + size` is U256 overflow, and it is also an error of return data out of bound):
```
let is_end_u256_overflow = sum.carry().as_ref().unwrap();
```

And rename other variables with `end` to `remainder_end`.

File: zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs (hunk 2 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs]: Updated this comment in [4ea268b](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1321/commits/4ea268b414ea757e452eab267e9a8a656c0cadcc).

@hero78119 may need help for review again. Thanks.

File: zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs (hunk 3 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs]: Fixed to `data_offset + size > U256::MAX` in [4ea268b](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1321/commits/4ea268b414ea757e452eab267e9a8a656c0cadcc).

File: zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs [hunk 4]
```diff
@@ -141,24 +149,27 @@ impl<F: Field> ExecutionGadget<F> for ErrorReturnDataOutOfBoundGadget<F> {
...
-            .assign(region, offset, F::from(end_overflow as u64))?;
+        let remainder_end_overflow = remainder_end.to_le_bytes()[N_BYTES_U64..]
+            .iter()
+            .fold(0, |acc, val| acc + u64::from(*val));
+        self.is_remainder_end_within_u64
+            .assign(region, offset, F::from(remainder_end_overflow))?;
 
         // check if it exceeds last callee return data length
-        let end_u64 = end.low_u64();
+        let reaminder_end_u64 = remainder_end.low_u64();
```
Comment (hero78119) [inline zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs:165]: there is typo 😃

**Review by hero78119 — COMMENTED** [review]
Find one typo, other all good!

File: zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs (hunk 4 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs]: Sorry for that. Fix it in [eb9a1c0](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1321/commits/eb9a1c0ba6d6de30caf72c613ff90753cd1ee1f3).

Comment (hero78119) [reply zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs]: no worry, all good now and tks!

**Pull request merged**